src/cpp/session/modules/SessionFind.cpp

```diff
@@ -301,26 +513,213 @@ class GrepOperation : public boost::enable_shared_from_this<GrepOperation>
+                     return error;
+                  }
+               }
+               if (findResults().preview())
+                  replaceString.insert(0, pLineInfo->decodedContents.substr(matchOn, matchSize));
+
+               Replacer replacer(findResults().ignoreCase());
+               if (findResults().regex() &&
+                   !findResults().replaceRegex())
+                  error = replacer.replaceRegexWithLiteral(matchOn, matchOff, &newLine, &searchPattern,
+                     &replacePattern, &replaceMatchOff);
+               else
+                  replacer.replaceLiteralWithLiteral(matchOn, matchOff, &newLine, &replaceString,
+                     &replaceMatchOff);
+               if (error)
+                  addErrorMessage(error.asString(), pErrorMessage,
+                     pReplaceMatchOn, pReplaceMatchOff, &lineSuccess);
+               if (lineSuccess)
+               {
+                  pLineInfo->decodedContents = newLine;
+
+                  // if multiple replaces in line, readjust previous match numbers
+                  if (pReplaceMatchOn->getSize() > 0 &&
+                      matchSize != replaceString.size())
+                  {
+                     json::Array tempMatchOn(*pReplaceMatchOn);
+                     json::Array tempMatchOff(*pReplaceMatchOff);
+                     pReplaceMatchOn->clear();
+                     pReplaceMatchOff->clear();
+
+                     // put back in reverse order
+                     int offset(gsl::narrow_cast<int>(replaceString.size() - matchSize));
+                     pReplaceMatchOn->push_back(json::Value(gsl::narrow_cast<int>(matchOn)));
+                     pReplaceMatchOff->push_back(json::Value(gsl::narrow_cast<int>(replaceMatchOff)));
+                     addOffsetIntegerToJsonArray(tempMatchOn, offset, pReplaceMatchOn);
+                     addOffsetIntegerToJsonArray(tempMatchOff, offset, pReplaceMatchOff);
+                  }
+                  else if (lineSuccess)
+                  {
+                     pReplaceMatchOn->push_back(json::Value(gsl::narrow_cast<int>(matchOn)));
+                     pReplaceMatchOff->push_back(json::Value(gsl::narrow_cast<int>(replaceMatchOff)));
+                  }
+               }
+               if (!findResults().preview())
+                  pProgress->addUnit();
+               pos--;
+            }
+            if (!findResults().preview())
+            {
+               Error error = encodeAndWriteLineToFile(pLineInfo->decodedContents,
+                  pLineInfo->leftContents, pLineInfo->rightContents);
+               if (error)
+                  addErrorMessage(error.asString(), pErrorMessage,
+                     pReplaceMatchOn, pReplaceMatchOff, &lineSuccess);
+            }
+         }
+      }
+      return Success();
+   }
+
    void onStdout(const core::system::ProcessOperations& ops, const std::string& data)
    {
       json::Array files;
```

Comment:
Can you clarify what you mean here?

Reply:
Instead of having 8 lines of delcaring empty json arrays you could have one or two lines by declaring them like:

```C++
json::Array files, lineNums, contents, matchOns, matchOffs, replaceMatchOns, replaceMatchOffs, errors; 
```

with an optional line break if you hit the line limit length. I'm not sure which I prefer though. I'd be interested to hear your opinion.